axelor-base/src/main/java/com/axelor/apps/base/service/UnitConversionServiceImpl.java

```diff
@@ -0,0 +1,218 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.base.service;
+
+import com.axelor.app.internal.AppFilter;
+import com.axelor.apps.base.AxelorException;
+import com.axelor.apps.base.db.Company;
+import com.axelor.apps.base.db.Product;
+import com.axelor.apps.base.db.Unit;
+import com.axelor.apps.base.db.UnitConversion;
+import com.axelor.apps.base.db.repo.TraceBackRepository;
+import com.axelor.apps.base.db.repo.UnitConversionRepository;
+import com.axelor.apps.base.exceptions.BaseExceptionMessage;
+import com.axelor.apps.base.service.app.AppBaseService;
+import com.axelor.apps.base.service.exception.TraceBackService;
+import com.axelor.auth.AuthUtils;
+import com.axelor.auth.db.User;
+import com.axelor.db.Model;
+import com.axelor.i18n.I18n;
+import com.axelor.utils.template.TemplateMaker;
+import com.google.inject.Inject;
+import groovy.lang.Binding;
+import groovy.lang.GroovyShell;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+import java.util.Optional;
+import org.codehaus.groovy.control.CompilationFailedException;
+import org.codehaus.groovy.control.CompilerConfiguration;
+import org.codehaus.groovy.control.customizers.ImportCustomizer;
+
+public class UnitConversionServiceImpl implements UnitConversionService {
+
+  private static final char TEMPLATE_DELIMITER = '$';
+  private static final int DEFAULT_COEFFICIENT_SCALE = 12;
+
+  protected AppBaseService appBaseService;
+
+  protected UnitConversionRepository unitConversionRepo;
+
+  @Inject
+  public UnitConversionServiceImpl(
+      AppBaseService appBaseService, UnitConversionRepository unitConversionRepo) {
+    this.appBaseService = appBaseService;
+    this.unitConversionRepo = unitConversionRepo;
+  }
+
+  /**
+   * Convert a value from a unit to another
+   *
+   * @param startUnit The starting unit
+   * @param endUnit The end unit
+   * @param value The value to convert
+   * @param scale The wanted scale of the result
+   * @param product Optional, a product used for complex conversions. Input null if needless.
+   * @return The converted value with the specified scale
+   * @throws AxelorException
+   */
+  @Override
+  public BigDecimal convert(
+      Unit startUnit, Unit endUnit, BigDecimal value, int scale, Product product)
+      throws AxelorException {
+    return genericConvert(startUnit, endUnit, value, scale, product, "Product");
+  }
+
+  protected BigDecimal genericConvert(
+      Unit startUnit, Unit endUnit, BigDecimal value, int scale, Model model, String nameInProject)
```

Comment:
Please rename this to "nameInContext" (to match the variable name in TemplateMaker.java)